contracts/test/UniswapMock.sol

```diff
@@ -0,0 +1,121 @@
+    }
+
+    function getToken(address exchange) external view returns (address token) {}
+    function getTokenWithId(uint256 tokenId) external view returns (address token) {}
+    // Never use
+    function initializeFactory(address template) external {}
+}
+
+
+contract UniswapExchangeMock is IUniswapExchange {
+    // Internal variables
+    address token;
+
+    constructor(address _token) public {
+        token = _token;
+    }
+
+    // Address of ERC20 token sold on this exchange
+    function tokenAddress() external view returns (address) {
+        return token;
+    }
+
+    // Address of Uniswap Factory
+    function factoryAddress() external view returns (address factory) {}
+
+    // Provide Liquidity
+    function addLiquidity(uint256 min_liquidity, uint256 max_tokens, uint256 deadline) external payable returns (uint256) {}
+    function removeLiquidity(uint256 amount, uint256 min_eth, uint256 min_tokens, uint256 deadline) external returns (uint256, uint256) {}
+
+    // Get Prices
+    function getEthToTokenInputPrice(uint256 eth_sold) external view returns (uint256 tokens_bought) {}
+    function getEthToTokenOutputPrice(uint256 tokens_bought) external view returns (uint256 eth_sold) {}
+    function getTokenToEthInputPrice(uint256 tokens_sold) external view returns (uint256 eth_bought) {}
+    function getTokenToEthOutputPrice(uint256 eth_bought) external view returns (uint256 tokens_sold) {}
+
+    // Trade ETH to ERC20
+    function ethToTokenSwapInput(uint256 min_tokens, uint256 deadline) external payable returns (uint256  tokens_bought) {
+        tokens_bought = msg.value * exchangeRates[ETH] / PPM;
+        ERC20(token).transfer(msg.sender, tokens_bought);
+    }
+
+    function ethToTokenTransferInput(uint256 min_tokens, uint256 deadline, address recipient) external payable returns (uint256  tokens_bought) {}
+    function ethToTokenSwapOutput(uint256 tokens_bought, uint256 deadline) external payable returns (uint256  eth_sold) {}
+    function ethToTokenTransferOutput(uint256 tokens_bought, uint256 deadline, address recipient) external payable returns (uint256  eth_sold) {}
+
+    // Trade ERC20 to ETH
+    function tokenToEthSwapInput(uint256 tokens_sold, uint256 min_eth, uint256 deadline) external returns (uint256  eth_bought) {}
+    function tokenToEthTransferInput(uint256 tokens_sold, uint256 min_eth, uint256 deadline, address recipient) external returns (uint256  eth_bought) {}
+    function tokenToEthSwapOutput(uint256 eth_bought, uint256 max_tokens, uint256 deadline) external returns (uint256  tokens_sold) {}
+    function tokenToEthTransferOutput(uint256 eth_bought, uint256 max_tokens, uint256 deadline, address recipient) external returns (uint256  tokens_sold) {}
+
+    // Trade ERC20 to ERC20
+    function tokenToTokenSwapInput(uint256 tokens_sold, uint256 min_tokens_bought, uint256 min_eth_bought, uint256 deadline, address token_addr) external returns (uint256  tokens_bought) {
```

Comment:
🚨 This is not how Uniswap works, see Vyper implementation: https://github.com/Uniswap/contracts-vyper/blob/master/contracts/uniswap_exchange.vy

In the actual implementation, it:
- Pulls `token` from `msg.sender`
- Convert `token` to ETH in this Exchange contract
- Call `Exchange(token_addr).ethToTokenTransferInput()` which transfers `token_addr` to `msg.sender`

I recommend using an actual implementation in Solidity instead of mocking it: https://github.com/PhABC/uniswap-solidity/blob/master/contracts/uniswap/UniswapExchange.sol